File: sdk/php/src/SigningIdentity.php
<?php

declare(strict_types=1);

namespace HideProtocol;

/**
 * A signing key. The seed stays inside the native library and is never exposed
 * to PHP — there is deliberately no accessor for it.
 *
 * Release it with close(); it is also released when the object is collected.
 */
final class SigningIdentity
{
    private ?\FFI\CData $handle;

    private function __construct(\FFI\CData $handle)
    {
        $this->handle = $handle;
    }

    /**
     * Creates an identity and returns the sealed key file to store.
     *
     * One seed backs both encryption and signing, so there is a single thing
     * to back up. A forgotten passphrase cannot be recovered: there is no escrow.
     */
    public static function generate(string $passphrase): string
    {
        if (strlen($passphrase) < Hide::MIN_PASSPHRASE_LEN) {
            throw new \InvalidArgumentException(
                'the passphrase must be at least ' . Hide::MIN_PASSPHRASE_LEN . ' characters'
            );
        }

        $out = Binding::emptyBuffer();
        $secret = Binding::cString($passphrase, 'passphrase');
        try {
            Binding::check(Binding::ffi()->hide_identity_generate($secret, \FFI::addr($out)));
        } finally {
            Binding::release($secret);
        }

        return Binding::take($out);
    }

    /**
     * Loads a signing identity from a key file.
     *
     * A key file written before signatures existed carries no signing seed and
     * is refused rather than being silently downgraded.
     */
    public static function load(string $bytes, ?string $passphrase = null): self
    {
        $ffi = Binding::ffi();
        $handle = $ffi->new('void *');
        $data = Binding::bytes($bytes);
        $secret = Binding::cString($passphrase, 'passphrase');

        try {
            Binding::check($ffi->hide_signing_identity_open(
                $ffi->cast('uint8_t *', $data),
                strlen($bytes),
                $secret,
                \FFI::addr($handle),
            ));
        } finally {
            Binding::release($data);
            Binding::release($secret);
        }

        return new self($handle);
    }

    /** The shareable verifying key, for others to check signatures with. */
    public function publicKey(): string
    {
        $handle = $this->alive();
        $out = Binding::emptyBuffer();
        Binding::check(Binding::ffi()->hide_signing_identity_public($handle, \FFI::addr($out)));

        return Binding::take($out);
    }

    /**
     * Signs $message under $context.
     *
     * The context separates uses of one identity, so a signature made for one
     * purpose cannot be presented as another. Never let a remote party choose it.
     */
    public function sign(string $context, string $message): string
    {
        $ffi = Binding::ffi();
        $handle = $this->alive();
        // Binary and length-prefixed: a context or message may contain a NUL.
        $contextBytes = Binding::bytes($context);
        $messageBytes = Binding::bytes($message);
        $out = Binding::emptyBuffer();

        try {
            Binding::check($ffi->hide_sign_message(
                $handle,
                $ffi->cast('uint8_t *', $contextBytes),
                strlen($context),
                $ffi->cast('uint8_t *', $messageBytes),
                strlen($message),
                \FFI::addr($out),
            ));
        } finally {
            Binding::release($contextBytes);
            Binding::release($messageBytes);
        }

        return Binding::take($out);
    }

    /** Answers a challenge, proving possession to whoever issued it. */
    public function answer(string $challenge): string
    {
        $ffi = Binding::ffi();
        $handle = $this->alive();
        $data = Binding::bytes($challenge);
        $out = Binding::emptyBuffer();

        try {
            Binding::check($ffi->hide_challenge_answer(
                $handle,
                $ffi->cast('uint8_t *', $data),
                strlen($challenge),
                \FFI::addr($out),
            ));
        } finally {
            Binding::release($data);
        }

        return Binding::take($out);
    }

    public function close(): void
    {
        if ($this->handle !== null) {
            Binding::ffi()->hide_signing_identity_free($this->handle);
            $this->handle = null;
        }
    }

    public function isClosed(): bool
    {
        return $this->handle === null;
    }

    /** @internal */
    public function alive(): \FFI\CData
    {
        if ($this->handle === null) {
            throw new \InvalidArgumentException('this identity has been closed');
        }

        return $this->handle;
    }

    public function __destruct()
    {
        $this->close();
    }

    /**
     * var_dump() and print_r() go through this, so it must never render key
     * material — not even a fingerprint derived from it.
     *
     * @return array{state: string}
     */
    public function __debugInfo(): array
    {
        return ['state' => $this->handle === null ? 'closed' : 'open'];
    }

    public function __toString(): string
    {
        return 'HideProtocol\SigningIdentity(' . ($this->handle === null ? 'closed' : 'open') . ')';
    }
}

File: sdk/dotnet/HideProtocol/SigningIdentity.cs
namespace HideProtocol;

/// <summary>
/// A signing key. The seed stays inside the native library and is never exposed
/// to managed code. Dispose it when finished.
/// </summary>
public sealed unsafe class SigningIdentity : IDisposable
{
    private IntPtr _handle;

    private SigningIdentity(IntPtr handle) => _handle = handle;

    /// <summary>
    /// Creates an identity, returning the sealed key file to store. One seed
    /// backs both encryption and signing, so there is a single thing to back
    /// up. A forgotten passphrase cannot be recovered: there is no escrow.
    /// </summary>
    public static byte[] Generate(string passphrase)
    {
        ArgumentNullException.ThrowIfNull(passphrase);
        if (passphrase.Length < Hide.MinPassphraseLength)
        {
            throw new ArgumentException(
                $"the passphrase must be at least {Hide.MinPassphraseLength} characters",
                nameof(passphrase));
        }

        using Utf8String pass = Utf8String.Create(passphrase, nameof(passphrase));
        HideBuffer output = Native.hide_buffer_empty();
        Interop.Check(Native.hide_identity_generate(pass.Pointer, &output));
        return Interop.Take(ref output);
    }

    /// <summary>
    /// Loads a signing identity. A key file written before signatures existed
    /// carries no signing seed and is refused rather than silently downgraded.
    /// </summary>
    public static SigningIdentity Load(ReadOnlySpan<byte> data, string? passphrase = null)
    {
        IntPtr handle = IntPtr.Zero;
        using Utf8String pass = Utf8String.Create(passphrase, nameof(passphrase));
        fixed (byte* pointer = data)
        {
            Interop.Check(Native.hide_signing_identity_open(
                pointer, (nuint)data.Length, pass.Pointer, &handle));
        }

        return new SigningIdentity(handle);
    }

    /// <summary>The shareable verifying key, for others to check signatures with.</summary>
    public byte[] PublicKey()
    {
        IntPtr handle = Alive();
        HideBuffer output = Native.hide_buffer_empty();
        Interop.Check(Native.hide_signing_identity_public(handle, &output));
        return Interop.Take(ref output);
    }

    /// <summary>
    /// Signs <paramref name="message"/> under <paramref name="context"/>, which
    /// separates uses of one identity. Never let a remote party choose it.
    /// </summary>
    public byte[] Sign(ReadOnlySpan<byte> context, ReadOnlySpan<byte> message)
    {
        IntPtr handle = Alive();
        HideBuffer output = Native.hide_buffer_empty();
        fixed (byte* contextPointer = context)
        fixed (byte* messagePointer = message)
        {
            Interop.Check(Native.hide_sign_message(
                handle,
                contextPointer,
                (nuint)context.Length,
                messagePointer,
                (nuint)message.Length,
                &output));
        }

        return Interop.Take(ref output);
    }

    /// <summary>Answers a challenge, proving possession to whoever issued it.</summary>
    public byte[] Answer(ReadOnlySpan<byte> challenge)
    {
        IntPtr handle = Alive();
        HideBuffer output = Native.hide_buffer_empty();
        fixed (byte* pointer = challenge)
        {
            Interop.Check(Native.hide_challenge_answer(
                handle, pointer, (nuint)challenge.Length, &output));
        }

        return Interop.Take(ref output);
    }

    private IntPtr Alive() =>
        _handle == IntPtr.Zero ? throw new ObjectDisposedException(nameof(SigningIdentity)) : _handle;

    public void Dispose()
    {
        if (_handle != IntPtr.Zero)
        {
            Native.hide_signing_identity_free(_handle);
            _handle = IntPtr.Zero;
        }

        GC.SuppressFinalize(this);
    }

    ~SigningIdentity() => Dispose();

    // Never render key material, not even a fingerprint of it.
    public override string ToString() =>
        $"HideProtocol.SigningIdentity ({(_handle == IntPtr.Zero ? "disposed" : "open")})";
}
